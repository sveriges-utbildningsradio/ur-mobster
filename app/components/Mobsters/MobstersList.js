// @flow

import React from 'react'
import { Droppable, Draggable } from 'react-beautiful-dnd'
import { FormattedMessage } from 'react-intl'
import type { User } from '../../types'
import styles from './Mobsters.css'
import avatarImage from '../../assets/avatar.png'
import removeButton from '../../assets/stop.png'

type MobstersListProps = {
  clickRemoveUser: (name: string, id: string) => void,
  droppableId: string,
  isEditing: boolean,
  users: User[]
}

const MobstersList = ({
  clickRemoveUser,
  droppableId,
  isEditing,
  users
}: MobstersListProps) => (
  <ul>
    <Droppable droppableId={droppableId}>
      {(provided, snapshot) => (
        /* eslint-disable react/jsx-props-no-spreading */
        <div
          {...provided.droppableProps}
          className={styles.droppableWrap}
          style={{
            backgroundColor: snapshot.isDraggingOver
              ? 'rgba(255, 255, 255, 0.1)'
              : 'rgb(17, 6, 54)'
          }}
          ref={provided.innerRef}
        >
          <ul data-e2e="mobsters-list-draggable-wrapper">
            {users && !!users.length ? (
              users.map((user, index) => (
                <Draggable
                  draggableId={user.name}
                  index={index}
                  key={user.id ? user.id : user.name}
                >
                  {providedDraggable => (
                    <li
                      className={styles.mobster}
                      key={user.id ? user.id : user.name}
                      ref={providedDraggable.innerRef}
                      {...providedDraggable.draggableProps}
                      {...providedDraggable.dragHandleProps}
                    >
                      <div className={styles.avatarWrap}>
                        <img
                          alt={user.name}
                          className={styles.avatar}
                          src={user.avatar ? user.avatar : avatarImage}
                        />
                        {index === 0 && droppableId === 'activeUsers' && (
                          <span className={styles.driverDot} />
                        )}
                        {index === 1 && droppableId === 'activeUsers' && (
                          <span className={styles.navigatorDot} />
                        )}
                      </div>
                      <div>
                        <p
                          className={styles.name}
                          data-e2e={`mobsterslist-name-${index}`}
                          data-testid={`mobsterslist-name-${index}`}
                        >
                          {user.name}
                        </p>
                        <p
                          className={styles.githubName}
                          data-e2e={`mobsterslist-githubName-${index}`}
                          data-testid={`mobsterslist-githubName-${index}`}
                        >
                          {user.githubName}
                        </p>
                      </div>

                      {isEditing && (
                        <input
                          alt={`Remove user from ${droppableId}`}
                          className={styles.editButton}
                          data-e2e="mobsterslist-remove-user"
                          data-testid={`mobsterslist-remove-${index}`}
                          onClick={() =>
                            clickRemoveUser(user.name, droppableId)
                          }
                          onKeyDown={e =>
                            e.keyCode === 13 &&
                            clickRemoveUser(user.name, droppableId)
                          }
                          src={removeButton}
                          type="image"
                        />
                      )}
                    </li>
                  )}
                </Draggable>
              ))
            ) : (
              <FormattedMessage id="noMobstersListItem" tagName="li" />
            )}
          </ul>
          {provided.placeholder}
        </div>
        /* eslint-enable react/jsx-props-no-spreading */
      )}
    </Droppable>
  </ul>
)

export default MobstersList
