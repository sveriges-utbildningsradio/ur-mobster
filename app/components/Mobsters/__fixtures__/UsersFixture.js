export const activeUsers = [
  {
    avatar: 'https://avatars1.githubusercontent.com/u/test',
    githubName: 'Active 1 Githubname',
    id: '2L_v21vEF',
    name: 'Active 1 Name'
  },
  { avatar: null, githubName: null, id: '9so97vODd', name: 'Active 2 Name' }
]

export const inactiveUsers = [
  {
    avatar: 'https://avatars1.githubusercontent.com/u/inactivetest',
    githubName: 'Inactive 1 Githubname',
    id: '2L_v21vEF',
    name: 'Inactive 2 Name'
  },
  { avatar: null, githubName: null, id: '9so97vODd', name: 'Inactive 2 Name' }
]
