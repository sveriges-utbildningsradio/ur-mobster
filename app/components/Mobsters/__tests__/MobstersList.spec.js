import React from 'react'
import { fireEvent, cleanup } from '@testing-library/react'
import 'jest-dom/extend-expect'
import renderWithReactIntl from '../../../utils/testHelpers'
import { DragDropContext } from 'react-beautiful-dnd'
import MobstersList from '../MobstersList'
import { activeUsers } from '../__fixtures__/UsersFixture'

describe('components/Mobsters/MobstersList', () => {
  let clickRemoveUser
  let users

  beforeEach(() => {
    clickRemoveUser = jest.fn()
  })

  afterEach(cleanup)

  it('should display a list of Mobsters', () => {
    const { getByTestId, container } = renderWithReactIntl(
      <DragDropContext onDragEnd={jest.fn()}>
        <MobstersList
          clickRemoveUser={clickRemoveUser}
          droppableId="testDroppable"
          isEditing={false}
          users={activeUsers}
        />
      </DragDropContext>
    )

    expect(getByTestId('mobsterslist-name-0')).toHaveTextContent(
      'Active 1 Name'
    )
    expect(getByTestId('mobsterslist-githubName-0')).toHaveTextContent(
      'Active 1 Githubname'
    )

    expect(getByTestId('mobsterslist-name-1')).toHaveTextContent(
      'Active 2 Name'
    )
    expect(getByTestId('mobsterslist-githubName-1')).toHaveTextContent('')

    expect(container.firstChild).toMatchSnapshot()
  })
})
