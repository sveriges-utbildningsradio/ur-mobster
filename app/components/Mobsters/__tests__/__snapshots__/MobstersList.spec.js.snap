// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Mobsters/MobstersList should display a list of Mobsters 1`] = `
<ul>
  <div
    class="droppableWrap"
    data-react-beautiful-dnd-droppable="0"
    style="background-color: rgb(17, 6, 54);"
  >
    <ul
      data-e2e="mobsters-list-draggable-wrapper"
    >
      <li
        aria-roledescription="Draggable item. Press space bar to lift"
        class="mobster"
        data-react-beautiful-dnd-drag-handle="0"
        data-react-beautiful-dnd-draggable="0"
        draggable="false"
        tabindex="0"
      >
        <div
          class="avatarWrap"
        >
          <img
            alt="Active 1 Name"
            class="avatar"
            src="https://avatars1.githubusercontent.com/u/test"
          />
        </div>
        <div>
          <p
            class="name"
            data-e2e="mobsterslist-name-0"
            data-testid="mobsterslist-name-0"
          >
            Active 1 Name
          </p>
          <p
            class="githubName"
            data-e2e="mobsterslist-githubName-0"
            data-testid="mobsterslist-githubName-0"
          >
            Active 1 Githubname
          </p>
        </div>
      </li>
      <li
        aria-roledescription="Draggable item. Press space bar to lift"
        class="mobster"
        data-react-beautiful-dnd-drag-handle="0"
        data-react-beautiful-dnd-draggable="0"
        draggable="false"
        tabindex="0"
      >
        <div
          class="avatarWrap"
        >
          <img
            alt="Active 2 Name"
            class="avatar"
            src="test-file-stub"
          />
        </div>
        <div>
          <p
            class="name"
            data-e2e="mobsterslist-name-1"
            data-testid="mobsterslist-name-1"
          >
            Active 2 Name
          </p>
          <p
            class="githubName"
            data-e2e="mobsterslist-githubName-1"
            data-testid="mobsterslist-githubName-1"
          />
        </div>
      </li>
    </ul>
  </div>
</ul>
`;
